qforce/forcefield.py

```diff
@@ -387,89 +394,187 @@ def set_charge(self, non_bonded):
+        for i_idx, (lj_type, a_name, q, mass) in enumerate(zip(non_bonded.lj_types,
+                                                               self.atom_names, self.q,
+                                                               self.masses), start=1):
+
+            mol2.write(f"{i_idx:8d} {lj_type} {coords[i_idx-1][0]:10.4f}"
+                       f"{coords[i_idx-1][1]:10.4f} {coords[i_idx-1][2]:10.4f}")
+            mol2.write(f" {unique_at[i_idx][atom_ids[i_idx]]} {1} {self.mol_name} {q:10.6f}\n")
+
+    def write_mol2_bond(self, mol2, topo, terms, atom_ids, unique_at):
+        n_bonds = 1
+        mol2.write("@<TRIPOS>BOND\n")
+        for bond in terms['bond']:
+            ids = bond.atomids + 1
+            mol2.write(f'{n_bonds:>6} {ids[0]:>6}{ids[1]:>6} un \n')
+            n_bonds = n_bonds + 1
+        mol2.write("@<TRIPOS>SUBSTRUCTURE\n")
+        mol2.write(f"{1:5} {self.mol_name} {1:8} TEMP {0:>8d} **** **** {0:5} ROOT\n")
+
+    def write_frcmod_mass(self, frcmod, non_bonded, atom_ids, unique_at):
+        frcmod.write(f'{self.mol_name} - frcmod generated by QForce\n')
+        frcmod.write('MASS\n')
+        for i, mass in enumerate((self.masses), start=1):
+            frcmod.write(f"{unique_at[i][atom_ids[i]]} {mass}\n")
+
+    def write_frcmod_bonds(self, frcmod, terms, alpha_map, atom_ids, unique_at):
+        frcmod.write("\nBOND\n")
+        for bond in terms['bond']:
+            ids = bond.atomids + 1
+            fconst = bond.fconst * (0.239005)/2  # kJ -> kcal
+            equ = bond.equ
+            frcmod.write(f"{unique_at[ids[0]][atom_ids[ids[0]]]:<2}-")
+            frcmod.write(f"{unique_at[ids[1]][atom_ids[ids[1]]]:<2}")
+            frcmod.write(f"{fconst:>10.2f}{equ:>10.2f}\n")
+
+    def write_frcmod_angles(self, frcmod, terms, atom_ids, unique_at):
+        frcmod.write("\nANGLE\n")
+        for angle in terms['angle']:
+            ids = angle.atomids + 1
+            fconst = angle.fconst * (0.239005)/2
+            equ = np.degrees(angle.equ)
+
+            if self.urey:
+                urey = [term for term in terms['urey'] if np.array_equal(term.atomids,
+                                                                         angle.atomids)]
+            if not self.urey or len(urey) == 0:
+                frcmod.write(f"{unique_at[ids[0]][atom_ids[ids[0]]]:<2}-")
+                frcmod.write(f"{unique_at[ids[1]][atom_ids[ids[1]]]:<2}-")
+                frcmod.write(f"{unique_at[ids[2]][atom_ids[ids[2]]]:<2}")
+                frcmod.write(f"{fconst:>10.2f}{equ:>10.2f}\n")
+            else:
+                urey_equ = urey[0].equ
+                urey_fconst = urey[0].fconst * (0.239005)/2
+                frcmod.write(f"{unique_at[ids[0]][atom_ids[ids[0]]]:<2}-")
+                frcmod.write(f"{unique_at[ids[1]][atom_ids[ids[1]]]:<2}-")
+                frcmod.write(f"{unique_at[ids[2]][atom_ids[ids[2]]]:<2}")
+                frcmod.write(f"{fconst:>10.2f}{equ:>10.2f}\n")
```

Comment:
@mateuszanotto, here you make variable urey_equ and urey_fconst, but then use fconst and equ instead. Is this a typo? Two options, either:
1)  urey_equ and urey_fconst are created unnecessarily (and should be removed in lines 483 484)
2) fconst and equ are typos and should be changed to urey_equ and urey_fconst in line 488

if it is option 2, I think all the qforce calculations you have run so far will have wrong urey bradley terms.

Let me know what you think or if im missing something

Reply:
Sorry for the delay. I was investigating it and AMBER doesn't support urey-bradley angles, so indeed they are created unnecessarily. I will remove lines 483 484.

Reply:
Removing those lines will not be a solution because Q-Force would still be fitting the FF assuming that you have UB terms on, then when you remove them, the resulting FF would be inaccurate.

UB term is simply an additional harmonic bond between atoms 1 and 3 of an angle, I am sure you could add this in AMBER.

Reply:
As far as I know, AMBER have only simple harmonic bond potential. It only support Urey-Bradley angle potential by converting CHARMM force field directly to AMBER topology file.

Reply:
@mateuszanotto , Urey-Bradley is simply a combination of two potentials:
1) A harmonic angle potential, which you already specify in the FF.
2 A harmonic bond potential, between atoms 1 and 3 of that angle

So all you have to do is adding "single harmonic bond potentials" between atoms 1 and 3 of your angles, using the force constants and equilibrium bond lengths that Q-Force has determined.

Exactly the same way you define standard bond potentials!

Reply:
Okay, I will add the bonds between atoms 1 and 3 and run the vib. modes/freq with it to check. 

As soon as I have them, I get back to you.

Thank you.

Reply:
Sounds great, please let me know if I can help somehow!